OpenRA.Mods.Common/Widgets/ObserverProductionIconsWidget.cs

```diff
@@ -132,5 +173,45 @@ public override Widget Clone()
 		{
 			return new ObserverProductionIconsWidget(this);
 		}
+
+		public override void MouseEntered()
+		{
+			if (TooltipContainer != null)
+				tooltipContainer.Value.SetTooltip(TooltipTemplate,
+					new WidgetArgs() { { "player", GetPlayer() }, { "getTooltipIcon", GetTooltipIcon } });
+		}
+
+		public override void MouseExited()
+		{
+			if (TooltipContainer == null)
+				return;
+
+			tooltipContainer.Value.RemoveTooltip();
+		}
+
+		public override void Tick()
+		{
+			for (var i = 0; i < iconRects.Length; i++)
```

Comment:
A small optimization would be to assume that the cursor hasn't moved since the last tick and check `iconRects[lastI].Contains(Viewport.LastMousePos)`.  If that's true then you don't need to loop over anything else.